Wrappers/Python/cil/optimisation/functions/SAGFunction.py

```diff
@@ -0,0 +1,187 @@
+
+    r"""
+    Stochastic average gradient (SAG) function
+    By incorporating a memory of previous gradient values the SAG method can achieve a faster convergence rate than black-box stochastic gradient methods. See the reference: Schmidt, M., Le Roux, N. and Bach, F., 2017. Minimizing finite sums with the stochastic average gradient. Mathematical Programming, 162, pp.83-112. https://doi.org/10.1007/s10107-016-1030-6. 
+
+    Parameters:
+    -----------
+    functions : `list`  of functions
+        A list of functions: :math:`[f_{0}, f_{2}, ..., f_{n-1}]`. Each function is assumed to be smooth with an implemented :func:`~Function.gradient` method. All functions must have the same domain. The number of functions (equivalently the length of the list) must be strictly greater than 1. 
+    sampler: An instance of a CIL Sampler class ( :meth:`~optimisation.utilities.sampler`) or of another class which has a `next` function implemented to output integers in :math:`{0,...,n-1}`.
+        This sampler is called each time `gradient` is called and sets the internal `function_num` passed to the `approximate_gradient` function.  Default is `Sampler.random_with_replacement(len(functions))`. 
+    
+    Note
+    ------
+    
+    The user has the option of calling the class method `warm_start_approximate_gradients` after initialising this class. This will compute and store the gradient for each function at an initial point. If this method is not called, the gradients are initialised with zeros. 
+
+
+    
+
+    """
+
+    def __init__(self, functions,  sampler=None):
+        self._list_stored_gradients = None
+        self._full_gradient_at_iterate = None
+        self._warm_start_just_done=False
+        self._sampled_grad=None
+        
+        super(SAGFunction, self).__init__(functions, sampler)
+
+
+        
+
+    def approximate_gradient(self, x, function_num,  out=None):
+        """ SAG approximate gradient, calculated at the point :math:`x` and updated using the function index given by `function_num`.  
+
+        Parameters
+        ----------
+        x: element in the domain of the `functions`
+
+        function_num: `int` 
+            Between 0 and the number of functions in the list  
+
+        """
+        
+        
+        if self._list_stored_gradients is None: # Initialise the stored gradients on the first call of gradient unless using warm start.  
+            self._list_stored_gradients = [
+                0*x for fi in self.functions]
+            self._full_gradient_at_iterate = 0*x
+            self._sampled_grad = x.copy()
+            self._stochastic_grad_difference = x.copy()
+        
+        if self.function_num >= self.num_functions or self.function_num<0 : #check the sampler and raise an error if needed
+            raise IndexError(
+                'The sampler has outputted an index larger than the number of functions to sample from. Please ensure your sampler samples from {0,1,...,len(functions)-1} only.')
```

Comment:
minor comment: "outputted" and "sampler samples" sound a bit odd. copilot suggests "The sampler has produced an index that exceeds the total number of available functions to sample from. Please make sure your sampler only selects from the set {0,1,...,len(functions)-1}." which sounds better (but can still be readjusted if needed)